[NCCL] create NCCL communicator for send/recv on demand

mingzhe09088 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45873 [NCCL] support send/recv to/from self when communicator is created on demand
* **#44922 [NCCL] create NCCL communicator for send/recv on demand**
* #44921 [NCCL] Support NCCL Send/Recv

For NCCL send/recv operations, we will create NCCL communicator on demand following the same design as how it's currently done for collective operations.

Differential Revision: [D23773726](https://our.internmc.facebook.com/intern/diff/D23773726/)

Commit:
[NCCL] create NCCL communicator for send/recv on demand

For NCCL send/recv operations, we will create NCCL communicator on demand following the same design as how it's currently done for collective operations.

Differential Revision: [D23773726](https://our.internmc.facebook.com/intern/diff/D23773726/)

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -718,7 +723,9 @@ void ProcessGroupNCCL::broadcastUniqueNCCLID(ncclUniqueId* ncclID) {
-    const std::vector<at::Device>& devices) {
+    const std::vector<at::Device>& devices,
+    NCCLCommType commType,
+    int myNewRank) {
```
nit: `p2pRank` might be a better name.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1078,8 +1097,11 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::pointToPoint(
       at::cuda::CUDAStream& ncclStream = ncclStreams_[key][i];
+      // For point-to-point communication, NCCL ranks can only
+      // be 0 or 1.
+      int newTargetRank = 1 - myNewRank;
```
myRank -> p2pRank, newTargetRank -> p2pTargetRank

mingzhe09088 commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -558,6 +571,13 @@ class ProcessGroupNCCL : public ProcessGroup {
+  // For point-to-point operations:
+  // The key is a string of my current rank and the peer process rank.
+  // e.g. If process 1 and process 2 are involved in a point-to-point communication,
+  // the key will be "1->2" on process 1 and "2->1" on process 2.
```
Not sure if I understand the question correctly. The current design only creates 1 communicator per pair of processes but the key is different. When process 1 send/recv to/from process 2, it will use the same communicator. We can sort the ranks so the key will be the same on both processes.

Commit:
Update on "[NCCL] create NCCL communicator for send/recv on demand"


For NCCL send/recv operations, we will create NCCL communicator on demand following the same design as how it's currently done for collective operations.

Differential Revision: [D23773726](https://our.internmc.facebook.com/intern/diff/D23773726/)

[ghstack-poisoned]

pritamdamania87 reviewed (approved):
LGTM! Although, I just realized that we should add a test to validate what happens when we send to self? For example, does NCCL support sending to rank 0 from rank 0? If so, we should add a test to validate our APIs support this correctly. If NCCL doesn't support this, we should add tests validating that our APIs raise an appropriate error.

mingzhe09088 commented:
NCCL supports sending to self. Let me add a test for that.

Commit:
Update on "[NCCL] create NCCL communicator for send/recv on demand"


For NCCL send/recv operations, we will create NCCL communicator on demand following the same design as how it's currently done for collective operations.

Differential Revision: [D23773726](https://our.internmc.facebook.com/intern/diff/D23773726/)

[ghstack-poisoned]

Commit:
Update on "[NCCL] create NCCL communicator for send/recv on demand"


For NCCL send/recv operations, we will create NCCL communicator on demand following the same design as how it's currently done for collective operations.

Differential Revision: [D23773726](https://our.internmc.facebook.com/intern/diff/D23773726/)

[ghstack-poisoned]

Commit:
Update on "[NCCL] create NCCL communicator for send/recv on demand"


For NCCL send/recv operations, we will create NCCL communicator on demand following the same design as how it's currently done for collective operations.

Differential Revision: [D23773726](https://our.internmc.facebook.com/intern/diff/D23773726/)

[ghstack-poisoned]
